### ISSUE TITLE
Calling for volunteers for developing cool features! 🚀

### ISSUE THREAD

**Issue by HAOCHENYE** [open-mmlab/mmengine#731]
![image](https://user-images.githubusercontent.com/57566630/202106190-7782900a-a17e-4530-bab5-402870e2900e.png)
# Introduction

Interested in deeply participating in OpenMMLab projects? Want to learn more about OpenMMLab's awesome tools without wasting plenty of time reading docs? The First OpenMMLab Codecamp has begun!! We provide more than a hundred tasks from seventeen research areas for you to choose from. Whether you are a novice in AI or a senior developer, there are suitable tasks for you to participate in. We will provide quick responses and full guidance to help you smoothly complete those tasks and upgrade to a core contributor of OpenMMLab. We cooperate with Beijing Super Cloud Center to provide computing power support. 

**How to participate?** 

Select the task you are interested in and submit registration [here](https://openmmlab.com/activity/codecamp/apply). We will inform you in three days if you have enrolled for the tasks, and then you can formulate the task plan with tutor and start development ! Once your PR has passed preliminary review, you can apply for the next task or just wait for the award!
More details: [OpenMMLab Activity page](https://openmmlab.com/activity/codecamp)。

# Task description

MMEngine, as the new generation training framework of OpenMMLab, will continue to absorb and integrate attractive cutting-edge technologies, however, our own power is limited, we hope to build an active community and leverage the power of the community to make MMEngine better and better. With the help of this event, we want to actively interact with the community to develop some very interesting features. We have graded some of these tasks as follows:

<table>
<thead>
<th>Task</th>
<th>Discription</th>
<th>Related skills</th>
<th>Difficulty</th>
<th>Credits</th>
<th>Status</t>
<body>
<tr>
  <td align="center">
  Support Segmind backend
  </td>
  <td align=left>
  Implement Sigmindbackend based on <a href="https://github.com/open-mmlab/mmengine/blob/f9158e233e0357798226254a1f57e0ad7d9e2491/mmengine/visualization/vis_backend.py#L59">BaseVisBackend</a>.

  **References**: <a href="https://github.com/open-mmlab/mmcv/blob/master/mmcv/runner/hooks/logger/segmind.py">SegmindLoggerHook</a>

  **step1**: Implement SegmindBackend
  
  **step2**: Update unit test

  **step3**: Update the visualization result in PR message

  </td>
  <td>
  Python, PyTorch
  </td>
  <td align="center">⭐️⭐️⭐️</td>
  <td align="center">10</td>
  <td align="center"> WIP </td>
</tr>
<tr>
  <td align="center">
  Support MLFlow backend
  </td>
  <td align=left>
  Implement MLFlowBackend based on <a href="https://github.com/open-mmlab/mmengine/blob/f9158e233e0357798226254a1f57e0ad7d9e2491/mmengine/visualization/vis_backend.py#L59">BaseVisBackend</a>

  **References**: <a href="https://github.com/open-mmlab/mmcv/blob/master/mmcv/runner/hooks/logger/mlflow.py">MlflowLoggerHook</a>

  **step1**: Implement MLFlowBackend
  
  **step2**: Update unit test

  **step3**: Update the visualization result in PR message

  </td>
  <td>
  Python, PyTorch
  </td>
  <td align="center">⭐️⭐️⭐️</td>
  <td align="center">10</td>
  <td align="center">WIP</td>
</tr>

<tr>
  <td align="center">
  Support Dvclive backend
  </td>
  <td align=left>
  Implement DvcliveBackend based on <a href="https://github.com/open-mmlab/mmengine/blob/f9158e233e0357798226254a1f57e0ad7d9e2491/mmengine/visualization/vis_backend.py#L59">BaseVisBackend</a>

  **References**: <a href="https://github.com/open-mmlab/mmcv/blob/master/mmcv/runner/hooks/logger/dvclive.py">DvcliveLoggerHook</a>

  **step1**: Implement DvcliveBackend
  
  **step2**: Update unit test

  **step3**: Update the visualization result in PR message

  </td>
  <td>
  Python, PyTorch
  </td>
  <td align="center">⭐️⭐️⭐️</td>
  <td align="center">10</td>
  <td align="center"></td>
</tr>

<tr>
  <td align="center">
  Support Neptune backend
  </td>
  <td align=left>
  Implement NeptuneBackend based on <a href="https://github.com/open-mmlab/mmengine/blob/f9158e233e0357798226254a1f57e0ad7d9e2491/mmengine/visualization/vis_backend.py#L59">BaseVisBackend</a>

  **References**: <a href="https://github.com/open-mmlab/mmcv/blob/master/mmcv/runner/hooks/logger/neptune.py">NeptuneBackend</a>

  **step1**: Implement NeptuneBackend
  
  **step2**: Update unit test

  **step3**: Update the visualization result in PR message

  </td>
  <td>
  Python, PyTorch
  </td>
  <td align="center">⭐️⭐️⭐️</td>
  <td align="center"></td>
  <td align="center">WIP</td>
</tr>

<tr>
  <td align="center">
  Support ClearML backend
  </td>
  <td align=left>
  Implement ClearMLBackend based on <a href="https://github.com/open-mmlab/mmengine/blob/f9158e233e0357798226254a1f57e0ad7d9e2491/mmengine/visualization/vis_backend.py#L59">BaseVisBackend</a>

  **References**: <a href="https://github.com/open-mmlab/mmcv/tree/master/mmcv/runner/hooks/logger">ClearMLLoggerHook</a>

  **step1**: Implement ClearMLBackend
  
  **step2**: Update unit test

  **step3**: Update the visualization result in PR message

  </td>
  <td>
  Python, PyTorch
  </td>
  <td align="center">⭐️⭐️⭐️</td>
  <td align="center">10</td>
  <td align="center"> WIP </td>
</tr>

<tr>
  <td align="center">
  Support ProfilerHook
  </td>
  <td align=left>
  Implement ProfilerHook based on <a href="https://github.com/open-mmlab/mmengine/blob/f9158e233e0357798226254a1f57e0ad7d9e2491/mmengine/hooks/hook.py#L7">Hook</a>

  **References**: <a href="https://github.com/open-mmlab/mmcv/blob/master/mmcv/runner/hooks/profiler.py">ProfilerHook</a>

  **step1**: Implement ProfilerHook
  
  **step2**: Update unit test

  **step3**: Update the effect in PR message

  </td>
  <td>
  Python, PyTorch
  </td>
  <td align="center">⭐️⭐️⭐️</td>
  <td align="center">10</td>
  <td align="center">WIP</td>
</tr>

<tr>
  <td align="center">
  Support early stop in MMEngine
  </td>
  <td align=left>
  Support stop training when validation metric reaches a threshold.

  **step1**: Design and implement early stop
  
  **step2**: Update unit test

  </td>
  <td>
  Python, PyTorch
  </td>
  <td align="center">⭐️⭐️⭐️</td>
  <td align="center">10</td>
  <td align="center"> <a href=https://github.com/open-mmlab/mmengine/pull/739>WIP</a></td>
</tr>

<tr>
  <td align="center">
  Support ApexOptimWrapper
  </td>
  <td align=left>
  Implement ApexOptimWrapper based on <a href="https://github.com/open-mmlab/mmengine/blob/f9158e233e0357798226254a1f57e0ad7d9e2491/mmengine/optim/optimizer/optimizer_wrapper.py#L16">OptimWrapper</a>

  **step1**: Implement ApexOptimWrapper
  
  **step2**: Update unit test

  </td>
  <td>
  Python, PyTorch
  </td>
  <td align="center">⭐️⭐️⭐️⭐️</td>
  <td align="center">30</td>
  <td align="center"><a href="https://github.com/open-mmlab/mmengine/pull/742">WIP</a></td>
</tr>

<tr>
  <td align="center">
  Support ReduceLROnPlateau
  </td>
  <td align=left>
  Implement ReduceLROnPlateau based on <a href="https://github.com/open-mmlab/mmengine/blob/f9158e233e0357798226254a1f57e0ad7d9e2491/mmengine/optim/scheduler/param_scheduler.py#L25">ParamScheduler</a>

  **References**: <a href="https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.ReduceLROnPlateau.html#torch.optim.lr_scheduler.ReduceLROnPlateau">torch.optim.ReduceLROnPlateau</a>

  **step1**: Implement ReduceLROnPlateau
  
  **step2**: Update unit test

  **step3**: Update the visualization of learning rate in PR message

  </td>
  <td>
  Python, PyTorch
  </td>
  <td align="center">⭐️⭐️⭐️⭐️</td>
  <td align="center">30</td>
  <td align="center">WIP</td>
</tr>

</thead>
</table>

Sign Up Here : [application form](https://openmmlab.com/activity/codecamp/apply) :rocket: :rocket: :rocket:  

By the way, we strongly encourage you to publish your experience on social media like medium or twitter with tag "OpenMMLab Codecamp" to share your experience with more developers! :laughing: :laughing:  :laughing:  

Discussion group: discord link

![image](https://user-images.githubusercontent.com/57566630/202130389-64567a18-7a4a-4240-8d9f-bd242ab93b93.png)

Welcome to join the discussion below or in discord. Come to take the challenge, and become a contributor to the OpenMMLab !
:partying_face: :partying_face:

**Comment by fcakyon** [comment]
Hello @HAOCHENYE, @zhouzaida, @hhaAndroid. Thanks for the great effort on all Open MMLab packages! We need your time to discuss some details.

@Blaizzy is from [Neptune.ai](https://neptune.ai/) and wants to discuss the roadmap of MMDetection and MMEngine with you. What questions do you have in mind @Blaizzy?

**Comment by Blaizzy** [comment]
Would you mind sharing when will MMDetection 3.x be released? And what are the main breaking changes from MMDetection 2.x?

**Comment by Blaizzy** [comment]
Additionally, I would like to tag my colleague @kshitij12345 who worked on the Neptune integration with the current MMDetection 2.x to chat. He will be also working on the new integration with MMEngine.

**Comment by kshitij12345** [comment]
Thanks for the intro @Blaizzy @fcakyon! 

Hello @HAOCHENYE, @zhouzaida, @hhaAndroid,

First of all, thank you for maintaining the great packages under the `open-mmlab` umbrella.
It would be great to discuss the road-map for packages under `open-mmlab` so that Neptune can seamlessly work with them. To that end, I was wondering if we should discuss this in a separate issue to keep this issue clean (or somewhere else?)

Thank you very much!

**Comment by HAOCHENYE** [comment]
Welcome @kshitij12345 @Blaizzy @fcakyon :santa::santa:!  Also, appreciate the fancy experiment track supported by Neptune :stuck_out_tongue_closed_eyes:, and we're very willing to collaborate with you to promote Neptune and OpenMMLab series repositories. I've opened a new [issue](#765) for discussing the roadmap, and we may shift there🚀
